# Class Data and Methods - Part 2

### 1. Use the class's name instead of self for the conn_count.

In [94]:
class ConnectionC:
    port = 55000
    conn_limit = 10
    conn_count = 0
    
    def __init__(self, host):
        # set the host for the instance
        self.host = host
        # set the port based on the class variable port
        
        # add 1 to the class's connection_count
        if ConnectionC.conn_count < ConnectionC.conn_limit:
            ConnectionC.conn_count += 1
        
        self.port = self.port + ConnectionC.conn_count

    def close(self):
        # reduce the class's connection_count by 1
        # self.conn_count -= 1
        ConnectionC.conn_count = ConnectionC.conn_count - 1
        
    def __repr__(self):
        return f"{self.host}, {self.port}"

In [95]:
connect = ConnectionC('localhost')
connect.__dict__

{'host': 'localhost', 'port': 55001}

In [99]:
connect.close()

In [100]:
connect.__dict__

{'host': 'localhost', 'port': 55001}

In [101]:
ConnectionC.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': -3,
              '__init__': <function __main__.ConnectionC.__init__(self, host)>,
              'close': <function __main__.ConnectionC.close(self)>,
              '__repr__': <function __main__.ConnectionC.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionC' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionC' objects>,
              '__doc__': None})

Each time I run connect.close(), it decrements conn_count. In this case, since I rand connect.close() three times in the notebook, it's set to -3.

In [102]:
connect2 = ConnectionC('localhost')
connect2.__dict__

{'host': 'localhost', 'port': 54998}

When using referencing conn_limit as a class variable, this works differently than if we had used it as an instance variable. 
Each time this is run, it increments the port by 1. But the odd side effect is that the port here is 54998, because I ran connect.close() more than once.

### 2. Working with subclasses

In [109]:
class ConnectionCa(ConnectionC):
    pass

I created a subclass of ConnectionCa here

In [110]:
connectCa1 = ConnectionCa('localhost')
connectCa1.__dict__

{'host': 'localhost', 'port': 54999}

That's quite odd ... it's incrementing the instance variable used in the previous class for the port. It does not start from port 55000.

In [111]:
ConnectionCa.__dict__

mappingproxy({'__module__': '__main__', '__doc__': None})

### 3. Using "__ class __" instead of the class's hardcoded name

In [103]:
class ConnectionD:
    port = 55000
    conn_limit = 10
    conn_count = 0
    
    def __init__(self, host):
        # set the host for the instance
        self.host = host
        # set the port based on the class variable port
        
        # add 1 to the class's connection_count
        if __class__.conn_count < __class__.conn_limit:
            __class__.conn_count += 1
        
        self.port = self.port + __class__.conn_count

    def close(self):
        # reduce the class's connection_count by 1
        # self.conn_count -= 1
        __class__.conn_count = __class__.conn_count - 1
        
    def __repr__(self):
        return f"{self.host}, {self.port}"

In [104]:
connectD = ConnectionD('localhost')
connectD.__dict__

{'host': 'localhost', 'port': 55001}

It seemed to work when we used __class__

In [105]:
connectD2 = ConnectionD('localhost')
connectD2.__dict__

{'host': 'localhost', 'port': 55002}

In [106]:
print(connectD)
print(connectD2)

localhost, 55001
localhost, 55002


In [108]:
ConnectionD.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': 2,
              '__init__': <function __main__.ConnectionD.__init__(self, host)>,
              'close': <function __main__.ConnectionD.close(self)>,
              '__repr__': <function __main__.ConnectionD.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionD' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionD' objects>,
              '__doc__': None})

### Using class methods (@classmethod decorator)

In [128]:
class ConnectionCM1:
    port = 55000
    conn_limit = 10
    conn_count = 0
    host = ''
    
    @classmethod
    def get_next_port(cls):
        return cls.port + 1
        
    @classmethod
    def add_connection(cls, host):
        cls.host = host
        
        # add 1 to the class's connection_count
        if cls.conn_count < cls.conn_limit:
            cls.conn_count += 1
        
        cls.port = cls.get_next_port()
        
    @classmethod
    def remove_connection(cls):
        # reduce the class's connection_count by 1
        cls.conn_count = cls.conn_count - 1
        
    @classmethod
    def get_connection_count(cls):
        return cls.conn_count
        

In [129]:
ConnectionCM1.add_connection('localhost')
ConnectionCM1.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55001,
              'conn_limit': 10,
              'conn_count': 1,
              'host': 'localhost',
              'get_next_port': <classmethod(<function ConnectionCM1.get_next_port at 0x108373420>)>,
              'add_connection': <classmethod(<function ConnectionCM1.add_connection at 0x1083737e0>)>,
              'remove_connection': <classmethod(<function ConnectionCM1.remove_connection at 0x108372520>)>,
              'get_connection_count': <classmethod(<function ConnectionCM1.get_connection_count at 0x1083736a0>)>,
              '__dict__': <attribute '__dict__' of 'ConnectionCM1' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionCM1' objects>,
              '__doc__': None})

In [131]:
ConnectionCM1.add_connection('localhost')
ConnectionCM1.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55002,
              'conn_limit': 10,
              'conn_count': 2,
              'host': 'localhost',
              'get_next_port': <classmethod(<function ConnectionCM1.get_next_port at 0x108373420>)>,
              'add_connection': <classmethod(<function ConnectionCM1.add_connection at 0x1083737e0>)>,
              'remove_connection': <classmethod(<function ConnectionCM1.remove_connection at 0x108372520>)>,
              'get_connection_count': <classmethod(<function ConnectionCM1.get_connection_count at 0x1083736a0>)>,
              '__dict__': <attribute '__dict__' of 'ConnectionCM1' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionCM1' objects>,
              '__doc__': None})

In [132]:
ConnectionCM1.remove_connection()
ConnectionCM1.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55002,
              'conn_limit': 10,
              'conn_count': 1,
              'host': 'localhost',
              'get_next_port': <classmethod(<function ConnectionCM1.get_next_port at 0x108373420>)>,
              'add_connection': <classmethod(<function ConnectionCM1.add_connection at 0x1083737e0>)>,
              'remove_connection': <classmethod(<function ConnectionCM1.remove_connection at 0x108372520>)>,
              'get_connection_count': <classmethod(<function ConnectionCM1.get_connection_count at 0x1083736a0>)>,
              '__dict__': <attribute '__dict__' of 'ConnectionCM1' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionCM1' objects>,
              '__doc__': None})

In [127]:
ConnectionCM.remove_connection()
ConnectionCM.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55006,
              'conn_limit': 10,
              'conn_count': 1,
              'host': 'localhost',
              'get_next_port': <classmethod(<function ConnectionCM.get_next_port at 0x1083725c0>)>,
              'add_connection': <classmethod(<function ConnectionCM.add_connection at 0x108373e20>)>,
              'remove_connection': <classmethod(<function ConnectionCM.remove_connection at 0x108373920>)>,
              'get_connection_count': <classmethod(<function ConnectionCM.get_connection_count at 0x108373b00>)>,
              '__dict__': <attribute '__dict__' of 'ConnectionCM' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionCM' objects>,
              '__doc__': None})